rfc/rfc-69/rfc-69.md

```diff
@@ -0,0 +1,159 @@
+
+## **State of the Project**
+
+As many of you know, Hudi was originally created at Uber in 2016 to solve [large-scale data ingestion](https://www.uber.com/blog/uber-big-data-platform/) and [incremental data processing](https://www.uber.com/blog/ubers-lakehouse-architecture/) problems and later [donated](https://www.uber.com/blog/apache-hudi/) to the ASF. 
+Since its graduation as a top-level Apache project in 2020, the community has made impressive progress toward the [streaming data lake vision](https://hudi.apache.org/blog/2021/07/21/streaming-data-lake-platform) to make data lakes more real-time and efficient with incremental processing on top of a robust set of platform components. 
+The most recent 0.13 brought together several notable features to empower incremental data pipelines, including - [_RFC-51 Change Data Capture_](https://github.com/apache/hudi/blob/master/rfc/rfc-51/rfc-51.md), more advanced indexing techniques like [_consistent hash indexes_](https://github.com/apache/hudi/blob/master/rfc/rfc-42/rfc-42.md) and 
+novel innovations like [_early conflict detection_](https://github.com/apache/hudi/blob/master/rfc/rfc-56/rfc-56.md) - to name a few.
+
+
+
+Today, Hudi [users](https://hudi.apache.org/powered-by) are able to solve end-end use cases using Hudi as a data lake platform that delivers a significant amount of automation on top of an interoperable open storage format. 
+Users can ingest incrementally from files/streaming systems/databases and insert/update/delete that data into Hudi tables, with a wide selection of performant indexes. 
+Thanks to the core design choices like record-level metadata and incremental/CDC queries, users are able to consistently chain the ingested data into downstream pipelines, with the help of strong stream processing support in 
+recent years in frameworks like Apache Spark, Apache Flink and Kafka Connect. Hudi's table services automatically kick in across this ingested and derived data to manage different aspects of table bookkeeping, metadata and storage layout. 
+Finally, Hudi's broad support for different catalogs and wide integration across various query engines mean Hudi tables can also be "batch" processed old-school style or accessed from interactive query engines.
+
+## **Future Opportunities**
+
+We're adding new capabilities in the 0.x release line, but we can also turn the core of Hudi into a more general-purpose database experience for the lake. As the first kid on the lakehouse block (we called it "transactional data lakes" or "streaming data lakes" 
```

Comment:
Generally speaking, the current implementation of Hudi is worked as micro-batch data lake, not a really streaming lakehouse. Will we propose to build the really streaming lakehouse via Hudi?

Reply:
This is a good point.  There was a discussion in 2021 on how we can make streaming writes more efficient, especially for Flink, by redesigning the core abstraction.  We should revisit that.  cc @danny0405 

https://lists.apache.org/thread/fsxbjm1w3gmn818lxn79lm6s56892s40

Reply:
cc @garyli1019 as he had a lot of ideas on this topic.

Reply:
@yihua, IMO, the proposal in above discussion is only about making the Flink writer in a streaming
fashion. But the streaming lakehouse kid is mainly end-to-end streaming read and write to make  the data in the lake really streaming processing.

Reply:
Good point! I think @SteNicholas is pointing towards more general-purpose streaming capabilities such as watermarks, windows and accumulators - https://www.oreilly.com/radar/the-world-beyond-batch-streaming-102/. Please correct me if i'm wrong.
We should certainly revive that devlist thread for a detailed discussion.

Reply:
@codope, reviving that devlist thread makes sense to me. But we could list this purpose for streaming lakehouse in this RFC.

Reply:
IMO "streaming" is a fancy word but vague in definition. If we referring to latency, it's impossible to do second-level latency on object store, maybe it's possible by adding cache(like alluxio?), but why the user not just use a HTAP database or message queue when they need such low latency. Referring stream like processing, what kind of stream processing the user will use when the latency is high. Materialized view could be a good fit with the lakehouse layer. Watermark is more on the ingestion layer, maybe we can add it to delta streamer. Happy to learn if there are more streaming use cases.

Reply:
@yihua feel like that discussion was long time ago lol. The reason I think we should refactor the core abstraction is that there are many unnecessary work to do when we want to support a new compute engine. If we keep the read and write core inside Hudi, adaptation will be easier and some performance improvement could be applied to all engines without additional work. Use writer as an example, if we have `HudiDriverClient` and `HudiWorkerClient` to handle all the table, timeline, file group and the actual write, support a new engine to ingest would be much easier, and the user shouldn't expect inconsistent behavior when they switch between Spark Flink Presto.

Reply:
+1 Materialized views are great and are supported in Trino as well. Should t be that hard to implement after Trino can write to Hudi.

Reply:
@vinothchandar, @kazdy, could we also support streaming view? In bilibili internal, we builds the streaming view on Hudi based on the watermark. Meanwhile, we also supports materialized views which is cached in Alluxio.

Reply:
> Could we not accomplish this with some enhancements in Hudi + Dynamic tables. cc @danny0405 thoughts

@vinothchandar Hudi + Dynamic table aims to resolve 2 use cases:

1. the ability to query the tables of intermediate layer, like the raw and silver tables, these tables can be updated in continuous streaming style and also queriable.
2. to unlock the streaming computation like aggregation in `REAL` incremeltal stype, the intermediate accumulation can be checkpointed and reused.

The end-to-end latency is in minutes range, not seconds or even milliseconds. I can see the materializatin view can be very useful in this streaming ingestion and near-real-time analytical use cases. The MVs can serve as a pre-aggregation layer to speed up the query raised by frontier users. If we make the pre-aggregation flexible enough, we can even embed custom agg logic from of user, to supply a direct service layer! Inspired by some engines like Apache Pino and Apache Kylin.

Reply:
In streaming world, watermark and event time are very important, no mater for query/write a single table or between multiple tables.
For example, query a snapshot of table at a specified event time instead of commit time. Or join snapshots of two table at the same event time. 
When building streaming warehouse or streaming materialized view, do we also consider introducing the concepts of event time and watermark in Hudi?

Reply:
@beyond1920 that's a good point of using watermark. +1 to provide the watermark to query engine. This could work pretty well with materialized view and some query logics are needed to get a view with precise eventtime.